src/main/java/edu/byu/cs/autograder/compile/ModuleIndependenceVerifier.java

```diff
@@ -5,53 +5,66 @@
 
 import java.io.File;
 import java.io.IOException;
-import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 
-public class ModuleIndependenceVerifier implements StudentCodeInteractor {
+public class ModuleIndependenceVerifier implements StudentCodeVerifier {
     @Override
-    public void interact(GradingContext context, StudentCodeReader reader) throws GradingException {
-        Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
-        Path serverPath = context.stageRepo().toPath().resolve("server/src/main/java");
-        Set<String> serverClasses = qualifiedNames(serverFiles, serverPath);
+    public void verify(GradingContext context, StudentCodeReader reader) throws GradingException {
+        try {
+            Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
+            Set<String> serverPackages = packageNames(serverFiles, reader);
 
-        Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
-        Path clientPath = context.stageRepo().toPath().resolve("client/src/main/java");
-        Set<String> clientClasses = qualifiedNames(clientFiles, clientPath);
+            Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
+            Set<String> clientPackages = packageNames(clientFiles, reader);
 
-        try {
-            checkImports(context, reader, serverFiles, clientClasses);
-            checkImports(context, reader, clientFiles, serverClasses);
+            checkImports(context, reader, serverFiles, clientPackages);
+            checkImports(context, reader, clientFiles, serverPackages);
         } catch (IOException e) {
             throw new GradingException("Unable to verify module independence", e);
         }
     }
 
-    private Set<String> qualifiedNames(Set<File> files, Path relativeTo) {
-        Set<String> classes = new HashSet<>();
+    private Set<String> packageNames(Set<File> files, StudentCodeReader reader) throws IOException {
+        Set<String> packages = new HashSet<>();
         for(File file : files) {
-            String fileName = relativeTo.relativize(file.toPath()).toString();
-            if(!fileName.contains(".java")) continue;
-            classes.add(fileName.substring(0, fileName.indexOf('.')).replace('/', '.'));
+            if(!file.getName().contains(".java")) continue;
+            List<String> contents = reader.getFileContents(file);
+            if(contents.isEmpty()) continue;
+            String firstLine = contents.getFirst();
+            if(firstLine.startsWith("package")) {
+                packages.add(firstLine.substring(8, firstLine.indexOf(';')));
```

Comment:
Is it possible for a student to create a **client package** named "MyPackage" and separately create a **server package** also named "MyPackage"? Could they do that with classes as well?

One approach would be to qualify our warnings by first detecting that it exists in the other module, and then excluding it if it is defined in the current module.

Reply:
Good point. How does it look now?

Reply:
Nice job handling those cases!